core/src/main/java/com/oath/oak/Slice.java

```diff
@@ -6,48 +6,164 @@
 class Slice {
-    private final int blockID;
-    private final ByteBuffer buffer;
-    // version with which this slice was allocated, if slice is not a result of a new creation the version can be invalid
-    private int version;
-    // This field is only used for sanity checks purposes and it should not be used, nor changed.
-    private final int originalPosition;
-
-    public Slice(int blockID, ByteBuffer buffer) {
+
+    protected int blockID;
+    protected int offset;
+    protected int length;
+
+    // Allocation time version
+    protected int version;
+
+    protected ByteBuffer buffer;
+
+    public Slice() {
+        invalidate();
+    }
+
+    // Used to duplicate the allocation state. Does not duplicate the buffer itself.
+    Slice(Slice otherSlice) {
+        copyFrom(otherSlice);
+    }
+
+    // Reset to invalid state
+    void invalidate() {
+        blockID = OakNativeMemoryAllocator.INVALID_BLOCK_ID;
+        offset = -1;
+        length = -1;
+        version = EntrySet.INVALID_VERSION;
+        buffer = null;
+    }
+
+    boolean isValid() {
+        return blockID != OakNativeMemoryAllocator.INVALID_BLOCK_ID;
+    }
+
+    /*
+     * Updates the allocation object.
+     * The buffer should be set later by the block allocator.
+     */
+    void update(int blockID, int offset, int length) {
         this.blockID = blockID;
+        this.offset = offset;
+        this.length = length;
+
+        // Invalidate the buffer and version. Will be assigned by the allocator.
+        this.version = EntrySet.INVALID_VERSION;
+        this.buffer = null;
+    }
+
+    // Set the internal buffer.
+    // This method should be used only by the block memory allocator.
+    void setBuffer(ByteBuffer buffer) {
         this.buffer = buffer;
-        this.originalPosition = buffer.position();
-        this.version = ValueUtilsImpl.INVALID_VERSION;
     }
 
-    Slice(int blockID, int position, int length, MemoryManager memoryManager) {
-        this(blockID, memoryManager.getByteBufferFromBlockID(blockID, position, length));
+    // Copy the block allocation information from another block allocation.
+    void copyFrom(Slice alloc) {
+        if (alloc == this) {
+            // No need to do anything if the input is this object
+            return;
+        }
+        this.blockID = alloc.blockID;
+        this.offset = alloc.offset;
+        this.length = alloc.length;
+        this.version = alloc.version;
+        this.buffer = alloc.buffer;
+    }
+
+    // Used by OffHeapList in "synchrobench" module, and for testings.
+    Slice duplicateBuffer() {
+        buffer = buffer.duplicate();
+        return this;
+    }
+
+    /* ------------------------------------------------------------------------------------
+     * Getters
+     * ------------------------------------------------------------------------------------*/
+
+    public int getAllocBlockID() {
```

Comment:
I renamed `Slice` methods:
 - Allocation info getters: `isAllocated()`, `getAllocatedBlockID()`, `getAllocatedOffset()`, `getAllocatedLength()`
   * used only by the allocator and codec
   * the exception is `isAllocated()` method that is used in many asserts in the code, but also in: `getValueState()`, `isKeyValid()`, and in `advance()`. For `advance()` we should consider using the `valueState` instead.
 - Metadata getters: `isValidVersion()`, `getVersion()`, `getMetadataAddress()`
   * `getMetadataAddress()` is used by `ValueUtilImpl` to update the metadata
 - Data getters (internal): `getDataByteBuffer()`, `getDataDuplicatedReadByteBuffer()`, `getDataDuplicatedWriteByteBuffer()`
   * used inetrnally by Oak to access the data
 - Data getters (OakUnsafeDirectBuffer): `getOffset()`, `getLength()`, `getAddress()`
   * used in `OakAttachedReadBuffer`
   * `OakAttachedReadBuffer` still need to implement its own `getByteBuffer()`:
      - `OakAttachedReadBuffer` will use `getDataDuplicatedReadByteBuffer()`
      - `OakAttachedWriteBuffer` will use `getDataDuplicatedWriteByteBuffer()`
      - Next PR will eliminate this as well because we will not need to duplicate the buffer anymore